tests/hkdf.cc

```diff
@@ -0,0 +1,239 @@
+/*
+ *    Copyright 2022 Cryspen Sarl
+ *
+ *    Licensed under the Apache License, Version 2.0 or MIT.
+ *    - http://www.apache.org/licenses/LICENSE-2.0
+ *    - http://opensource.org/licenses/MIT
+ */
+
+#include <fstream>
+
+#include <gtest/gtest.h>
+#include <nlohmann/json.hpp>
+
+#include "Hacl_HKDF.h"
+#include "Hacl_Spec.h"
+#include "util.h"
+
+using json = nlohmann::json;
+
+typedef struct
+{
+  std::string comment;
+  // A tcId of 0 indicates "not Wycheproof".
+  uint32_t tcId;
+  std::string hash;
+  // 1) Input Keying Material
+  bytes IKM;
+  bytes salt;
+  bytes info;
+  // Length of OKM
+  uint32_t L;
+  // 2) A pseudorandom key (of HashLen octets)
+  // Only available when "not Wycheproof".
+  bytes PRK;
+  // 3) Output Keying Material
+  bytes OKM;
+  bool valid;
+} TestCase;
+
+class HKDFSuite : public ::testing::TestWithParam<TestCase>
+{};
+
+TEST_P(HKDFSuite, TestCase)
+{
+  auto test = GetParam();
+  std::cout << "# " << test.comment << std::endl;
+
+  // TODO: How does error handling in Hacl_HKDF_* work?
+  if (!test.valid) {
+    return;
+  }
+
+  if (test.hash == "SHA-1") {
+    std::cout << "Skipping... (SHA-1 not supported)" << std::endl;
+  } else if (test.hash == "SHA-256") {
+    bytes got_prk = std::vector<uint8_t>(32);
+
+    Hacl_HKDF_extract_sha2_256(got_prk.data(),
+                               test.salt.data(),
+                               test.salt.size(),
+                               test.IKM.data(),
+                               test.IKM.size());
+
+    if (test.tcId == 0) {
+      EXPECT_EQ(test.PRK, got_prk);
+    }
+
+    bytes got_OKM = std::vector<uint8_t>(test.L);
+
+    Hacl_HKDF_expand_sha2_256(got_OKM.data(),
+                              got_prk.data(),
+                              got_prk.size(),
+                              test.info.data(),
+                              test.info.size(),
+                              test.L);
+
+    EXPECT_EQ(test.OKM, got_OKM);
+  } else if (test.hash == "SHA-512") {
+    bytes got_prk = std::vector<uint8_t>(64);
+
+    Hacl_HKDF_extract_sha2_512(got_prk.data(),
+                               test.salt.data(),
+                               test.salt.size(),
+                               test.IKM.data(),
+                               test.IKM.size());
+
+    if (test.tcId == 0) {
+      EXPECT_EQ(test.PRK, got_prk);
+    }
+
+    bytes got_OKM = std::vector<uint8_t>(test.L);
+
+    Hacl_HKDF_expand_sha2_512(got_OKM.data(),
+                              got_prk.data(),
+                              got_prk.size(),
+                              test.info.data(),
+                              test.info.size(),
+                              test.L);
+
+    EXPECT_EQ(test.OKM, got_OKM);
+  } else if (test.hash == "blake2s") {
+    FAIL()
+      << "This should not run. We have no test vectors for HKDF-BLAKE2S yet.";
```

Comment:
You could generate a few tests with something like https://cryptography.io/en/3.4.2/development/custom-vectors/hkdf.html to have at least something here.

Reply:
Will do, thanks.